include/sympack/cuBLAS.hpp

```diff
@@ -0,0 +1,627 @@
+/* cuBLAS interface */
+#include  "sympack/Environment.hpp"
+#ifdef CUDA_MODE
+#include "cuda_runtime.h"
+#include "device_launch_parameters.h"
+#include "cublas_v2.h"
+
+                                                                                               \
+
+namespace symPACK {
+namespace cublas {
+    // NOTE: This only contains a subset of the complete BLAS, just the operations used in symPACK are included.
+    // SYRK, TRSM, GEMM, Axpy, Scal, Gemv, Geru, Copy
+    // A lot of these operations are only used in the defunct SuperNode methods that have been replaced with
+    // the methods in symPACKMatrix2D.hpp, but they're here just in case they become needed at some point.
+    typedef  int                    Int;
+    /* ===== LEVEL 1 BLAS ===== */
+
+    /* AXPY */
+    cublasStatus_t cublas_axpy(cublasHandle_t, Int,
+                           const float           *,
+                           const float           *, Int,
+                           float                 *, Int);
+
+    cublasStatus_t cublas_axpy(cublasHandle_t, Int,
+                           const double           *,
+                           const double           *, Int,
+                           double                 *, Int);
+
+    cublasStatus_t cublas_axpy(cublasHandle_t, Int,
+                           const cuComplex           *,
+                           const cuComplex           *, Int,
+                           cuComplex                 *, Int);
+
+    cublasStatus_t cublas_axpy(cublasHandle_t, Int,
+                           const cuDoubleComplex           *,
+                           const cuDoubleComplex           *, Int,
+                           cuDoubleComplex                 *, Int);
+
+    template <typename T>
+    cublasStatus_t cublas_axpy_wrapper(Int N,
+                           const T           DA,
+                           const T           * DX, Int incx,
+                           T                 * DY, Int incy) {
+        
+        int rank;
+        int gpu_id;
+        int n_gpus;
+
+        MPI_Comm_rank(symPACK::world_comm, &rank);
+        cudaGetDeviceCount(&n_gpus);
+
+        gpu_id = rank % n_gpus;
+
+        cudaSetDevice(gpu_id);
```

Comment:
IIUC the lines above (and similar lines duplicated in template functions below) are selecting a local GPU based on the _global_ rank of this process. This happens to have reasonable behavior for job layout where tasks-per-node happen to be a multiple of gpus-per-node, but I'm pretty sure it "does the wrong thing" for job layouts lacking this property (e.g. 1 tasks-per-node and 2 GPUs-per-node). 

It would make slightly more sense to use the _local_ rank of the process (ie. `upcxx::local_team().rank_me()`), but would be even better to just use `symPACK::gpu_allocator.device_id()`:

```suggestion
        cudaSetDevice(symPACK::gpu_allocator.device_id());
```

Wider question: `symPACK_cuda_setup()` appears to already do exactly that, so **why are these `cudaSetDevice()` calls required at all?**

Reply:
See answer to comment below, this is dead code that never should have been in the PR. Sorry for the confusion caused by this.